### ISSUE TITLE
Not able to push (for long time) if the buildpack gets updated during the staging process

### ISSUE THREAD

**Issue by krumts** [cloudfoundry/cloud_controller_ng#1362]
## Issue

We had the issue described below recently with one of our productive apps, and now I got a well reproducible scenario.
If I push an app with with a specific buildpack and in parallel the buildpack gets updated (with 'cf update-buildpack') while the staging process is running, then:
- on the client side I see staging timeout after 15 min and 
- subsequent attempts to push the app fail for long periods of time 

## Context
I'll mix the detailed explanation with the steps to reproduce

## Steps to Reproduce

- Downloaded 2 versions of a buildpack - I tested with a home-grown buildpack and with the standard node.js one - 1.6.50 and 1.6.51
- created a new buildpack entry for the node.js one:
```cf create-buildpack test_node_buildpack nodejs-buildpack-cflinuxfs3-v1.6.50.zip 16```
- push an app with the buildpack (this works fine)
```cf push -b test_node_buildpack```
- As a next step, try in parallel (two separate shells) to:
  - ```cf push -b test_node_buildpack```
  - ```cf update-buildpack test_node_buildpack -p nodejs-buildpack-cflinuxfs3-v1.6.51.zip 16```
  - the issue happens only when 
    - the old version of the buildpack is downloaded and used for staging (see 1.50.0 in the logs below)
    - the buildpack gets updated _before_ staging finishes
  - it may take a few attempts to get the correct timing to reproduce the issue 
- the effect is that the staging process gets stuck for 15 minutes and ends with an timeout error:
```
$ cf push -b test_node_buildpack
Pushing from manifest to org test / space sample_apps as admin...
Using manifest file /Users/myuser/git/cf-sample-apps/cf-example-env/manifest.yml
Getting app info...
Updating app with these attributes...
  name:                nodejsexample
  path:                /Users/myuser/git/cf-sample-apps/cf-example-env
  buildpacks:
    test_node_buildpack
  command:             node app.js
  disk quota:          1G
  health check type:   port
  instances:           1
  memory:              128M
  stack:               cflinuxfs3
  routes:
    nodejsexample.cfapps.dev22.aws.maskeddomain.io

Updating app nodejsexample...
Mapping routes...
Comparing local files to remote cache...
Packaging files to upload...
Uploading files...
 938 B / 938 B [==================================================================================] 100.00% 1s

Waiting for API to complete processing files...

Stopping app...

Staging app and tracing logs...
   Downloading test_node_buildpack...
   Downloaded test_node_buildpack
   Cell 2d8b75a1-de5a-476f-9379-728c7b029545 creating container for instance b8438bda-54c6-429d-8e10-20a8b5df063d
   Cell 2d8b75a1-de5a-476f-9379-728c7b029545 successfully created container for instance b8438bda-54c6-429d-8e10-20a8b5df063d
   Downloading app package...
   Downloading build artifacts cache...
   Downloaded app package (938B)
   Downloaded build artifacts cache (19.7M)
   -----> Nodejs Buildpack version 1.6.50
   -----> Installing binaries
          engines.node (package.json): unspecified
          engines.npm (package.json): unspecified (use default)
          **WARNING** Node version not specified in package.json or .nvmrc. See: http://docs.cloudfoundry.org/buildpacks/node/node-tips.html
   -----> Installing node 10.15.3
          Copy [/tmp/cache/final/dependencies/627b7fd2027abd42c8d93234ec6209e0272366485933a502c397c18aea33582a/node-10.15.3-linux-x64-cflinuxfs3-52dbc618.tgz]
          Using default npm version: 6.4.1
   -----> Installing yarn 1.15.2
          Copy [/tmp/cache/final/dependencies/f0b0c3270dff2f11d91915a30302f19f6e252fcd55c95188bcedbd532c7412e2/yarn-1.15.2-any-stack-c4feca9b.tar.gz]
          Installed yarn 1.15.2
   -----> Creating runtime environment
          PRO TIP: It is recommended to vendor the application's Node.js dependencies
          Visit http://docs.cloudfoundry.org/buildpacks/node/index.html#vendoring
          NODE_ENV=production
          NODE_HOME=/tmp/contents954563519/deps/0/node
          NODE_MODULES_CACHE=true
          NODE_VERBOSE=false
          NPM_CONFIG_LOGLEVEL=error
          NPM_CONFIG_PRODUCTION=true
   -----> Building dependencies
          Installing node modules (package.json)
   up to date in 0.518s
   found 0 vulnerabilities
   Exit status 0
   Uploading droplet, build artifacts cache...
   Uploading droplet...
   Uploading build artifacts cache...
   Uploaded build artifacts cache (19.7M)
   Uploaded droplet (19.9M)
   Uploading complete
   Cell 2d8b75a1-de5a-476f-9379-728c7b029545 stopping instance b8438bda-54c6-429d-8e10-20a8b5df063d
   Cell 2d8b75a1-de5a-476f-9379-728c7b029545 destroying container for instance b8438bda-54c6-429d-8e10-20a8b5df063d
   Cell 2d8b75a1-de5a-476f-9379-728c7b029545 successfully destroyed container for instance b8438bda-54c6-429d-8e10-20a8b5df063d
Error staging application nodejsexample: timed out after 15 minute(s)
FAILED
```
- looking through the logs, I see that the cloud_controller logged an error:
```diego.staging.buildpack.saving-staging-result-failed``` with an error message:
```"error":"Must specify either a buildpack_url or an admin_buildpack_name"```
Here is the full log message:
```
<14>1 2019-06-04T13:13:24.96681Z 10.0.129.68 cloud_controller_ng rs2 - [instance@47450 director="bosh-dev22" deployment="cf" group="api" az="z2" id="0ac3ed3d-3dcf-463f-9714-51df548d29bd"] {"timestamp":1559654004.4061115,"message":"diego.staging.buildpack.saving-staging-result-failed","log_level":"error","source":"cc.stager","data":{"request_guid":"117482f8-0739-46d9-81a0-cd7b84f6fe82","staging_guid":"b8438bda-54c6-429d-8e10-20a8b5df063d","response":{"result":{"lifecycle_metadata":{"buildpack_key":"5e35e5bc-e08b-49ab-8312-cc172cebd806_cd13fd76adfe19808cb6fc3b9eee8af1d52cc0df311b659bff8116239367dbad","detected_buildpack":"nodejs","buildpacks":[{"key":"5e35e5bc-e08b-49ab-8312-cc172cebd806_cd13fd76adfe19808cb6fc3b9eee8af1d52cc0df311b659bff8116239367dbad","name":"nodejs","version":"1.6.50"}]},"process_types":{"web":"npm start"},"execution_metadata":"","lifecycle_type":"buildpack"}},"error":"Must specify either a buildpack_url or an admin_buildpack_name","build":"#<VCAP::CloudController::BuildModel @values={:id=>49381, :guid=>\"b8438bda-54c6-429d-8e10-20a8b5df063d\", :created_at=>2019-06-04 13:13:11 UTC, :updated_at=>2019-06-04 13:13:11 UTC, :state=>\"STAGING\", :package_guid=>\"1f06ac45-798b-4a1f-86c0-58105fa27dbd\", :error_description=>nil, :app_guid=>\"3cbaa9d8-a90d-43ef-92a6-331e89a359cd\", :error_id=>nil, :created_by_user_guid=>\"a613649f-7c73-4444-b185-ea678dddf6d3\", :created_by_user_name=>\"admin\", :created_by_user_email=>\"admin\"}>","droplet":"#<VCAP::CloudController::DropletModel @values={:id=>42517, :guid=>\"f4ebc3da-f41a-4524-aa95-fd8ea980db5e\", :created_at=>2019-06-04 13:13:20 UTC, :updated_at=>2019-06-04 13:13:22 UTC, :droplet_hash=>\"ff3e1b16b2643b9ea3bcce56db7f5b924836eabe\", :execution_metadata=>nil, :state=>\"STAGING\", :process_types=>nil, :error_id=>nil, :error_description=>nil, :encrypted_environment_variables=>nil, :salt=>nil, :staging_memory_in_mb=>nil, :staging_disk_in_mb=>nil, :buildpack_receipt_buildpack=>\"test_node_buildpack\", :buildpack_receipt_buildpack_guid=>nil, :buildpack_receipt_detect_output=>\"nodejs\", :docker_receipt_image=>nil, :package_guid=>\"1f06ac45-798b-4a1f-86c0-58105fa27dbd\", :app_guid=>\"3cbaa9d8-a90d-43ef-92a6-331e89a359cd\", :sha256_checksum=>\"b6fe3addda3a7848ed63aff6f2d49ae463a1979b85c3b993634be7da310ab589\", :build_guid=>\"b8438bda-54c6-429d-8e10-20a8b5df063d\", :docker_receipt_username=>nil, :docker_receipt_password_salt=>nil, :encrypted_docker_receipt_password=>nil, :encryption_key_label=>\"encryption_key_0\", :encryption_iterations=>2048}>"},"thread_id":47269241043200,"fiber_id":47269224244420,"process_id":1,"file":"/var/vcap/data/packages/cloud_controller_ng/ff1c2e10f33ff378963ecd66182d2baa150cb0e8/cloud_controller_ng/lib/cloud_controller/diego/staging_completion_handler.rb","lineno":103,"method":"rescue in handle_success"}
```

- retry the cf push command:
```
$ date
Tue Jun  4 15:35:05 CEST 2019
$ cf push -b test_node_buildpack
Pushing from manifest to org test / space sample_apps as admin...
Using manifest file /Users/myuser/git/cf-sample-apps/cf-example-env/manifest.yml
Getting app info...
Updating app with these attributes...
  name:                nodejsexample
  path:                /Users/myuser/git/cf-sample-apps/cf-example-env
  buildpacks:
    test_node_buildpack
  command:             node app.js
  disk quota:          1G
  health check type:   port
  instances:           1
  memory:              128M
  stack:               cflinuxfs3
  routes:
    nodejsexample.cfapps.dev22.aws.maskeddomain.io

Updating app nodejsexample...
Mapping routes...
Comparing local files to remote cache...
Packaging files to upload...
Uploading files...
 938 B / 938 B [==================================================================================] 100.00% 1s

Waiting for API to complete processing files...

Stopping app...

Staging app and tracing logs...
The app is invalid: VCAP::CloudController::BuildCreate::StagingInProgress
FAILED
```

- only about 20 minutes after the ```Must specify either a buildpack_url or an admin_buildpack_name``` a cc.worker logs:
``` cc-worker	Staging timeout has elapsed for build: b8438bda-54c6-429d-8e10-20a8b5df063d```
- After seeing this error a few times, I start getting a different one - Error staging application: StagingTimeExpired 

```
$ date
Tue Jun  4 15:37:20 CEST 2019
$ cf push -b test_node_buildpack
Pushing from manifest to org test / space sample_apps as admin...
Using manifest file /Users/myuser/git/cf-sample-apps/cf-example-env/manifest.yml
Getting app info...
Updating app with these attributes...
  name:                nodejsexample
  path:                /Users/myuser/git/cf-sample-apps/cf-example-env
  buildpacks:
    test_node_buildpack
  command:             node app.js
  disk quota:          1G
  health check type:   port
  instances:           1
  memory:              128M
  stack:               cflinuxfs3
  routes:
    nodejsexample.cfapps.dev22.aws.maskeddomain.io

Updating app nodejsexample...
Mapping routes...
Comparing local files to remote cache...
Packaging files to upload...
Uploading files...
 938 B / 938 B [==================================================================================] 100.00% 1s

Waiting for API to complete processing files...

Stopping app...

Staging app and tracing logs...
Error staging application: StagingTimeExpired
FAILED
```
- after some time the push works again - if I properly analyzed the logs this happens after the old binaries for the buildpack get deleted - at least I've seen the delete operations in the bits service logs

## Expected result

I would expect that the app push finishes successfully with the buildpack downloaded during staging
In the worst case I would expect one failed push and that next push works

## Current result

I see staging timeout on the client first after 15 min
```
Uploaded droplet (19.9M)
Uploading complete
Cell 2d8b75a1-de5a-476f-9379-728c7b029545 stopping instance b8438bda-54c6-429d-8e10-20a8b5df063d
Cell 2d8b75a1-de5a-476f-9379-728c7b029545 destroying container for instance b8438bda-54c6-429d-8e10-20a8b5df063d
Cell 2d8b75a1-de5a-476f-9379-728c7b029545 successfully destroyed container for instance b8438bda-54c6-429d-8e10-20a8b5df063d
Error staging application nodejsexample: timed out after 15 minute(s)
FAILED
```
Retried attempts to push the app after the error also fail:
```
Waiting for API to complete processing files...

Stopping app...

Staging app and tracing logs...
The app is invalid: VCAP::CloudController::BuildCreate::StagingInProgress
FAILED
```
and later on:
```
Error staging application: StagingTimeExpired
FAILED
```

## Possible Fix

Can't suggest. I also haven't fully understood why the issue happens, but hope that the steps to reproduce above would help

**Comment by karlismelderis** [comment]
I was able to reproduce same behaviour by specifying higher NPM version in package.json than default in build pack.

**Comment by cwlbraa** [comment]
@karlismelderis that is not the same issue. please open an issue on the npm buildpack.

**Comment by cwlbraa** [comment]
This does some like a legit race, but it's hard to see how we'd fix it. I suspect either the push is getting mixed signals (should it use the buildpack from before the update or after?) OR it's just waiting for updated buildpack to flow all the way through the pipes (client->api->local-worker->blobstore->staging container) and timing out in the process.

Unless this is plaguing your environment, I'm going to close this as wontfix. Please feel free to reopen if you need to.

**Comment by ChrisChross** [comment]
Any solutions on this? I get the same error message. I also tried to set -t param but it says max. 600s possible.
